pkg/networkservice/common/dial/client.go

```diff
@@ -0,0 +1,111 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dial will dial up a grpc.ClientConnInterface if a client *url.URL is provided in the ctx, retrievable by
+// clienturlctx.ClientURL(ctx) and put the resulting grpc.ClientConnInterface into the ctx using clientconn.Store(..)
+// where it can be retrieved by other chain elements using clientconn.Load(...)
+package dial
+
+import (
+	"context"
+	"time"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/postpone"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clientconn"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
+)
+
+type dialClient struct {
+	chainCtx    context.Context
+	dialOptions []grpc.DialOption
+	dialTimeout time.Duration
+}
+
+// NewClient - returns new dial chain element
+func NewClient(chainCtx context.Context, opts ...Option) networkservice.NetworkServiceClient {
+	o := &option{}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return &dialClient{
+		chainCtx:    chainCtx,
+		dialOptions: o.dialOptions,
+		dialTimeout: o.dialTimeout,
+	}
+}
+
+func (d *dialClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	closeContextFunc := postpone.ContextWithValues(ctx)
+	// If no clientURL, we have no work to do
+	// call the next in the chain
+	clientURL := clienturlctx.ClientURL(ctx)
+	if clientURL == nil {
+		return next.Client(ctx).Request(ctx, request, opts...)
+	}
+
+	cc, _ := clientconn.LoadOrStore(ctx, newDialer(d.chainCtx, d.dialTimeout, d.dialOptions...))
+
+	// If there's an existing grpc.ClientConnInterface and it's not ours, call the next in the chain
+	di, ok := cc.(*dialer)
+	if !ok {
+		return next.Client(ctx).Request(ctx, request, opts...)
+	}
+
+	// If our existing dialer has a different URL close down the chain
+	if di.clientURL != nil && di.clientURL.String() != clientURL.String() {
+		closeCtx, closeCancel := closeContextFunc()
+		defer closeCancel()
+		_ = di.Dial(closeCtx, di.clientURL)
+		_, _ = next.Client(ctx).Close(clienturlctx.WithClientURL(closeCtx, di.clientURL), request.GetConnection(), opts...)
+	}
+
+	err := di.Dial(ctx, clientURL)
+	if err != nil {
+		clientconn.Delete(ctx)
+		closeCtx, closeCancel := closeContextFunc()
+		defer closeCancel()
+		_, _ = next.Client(ctx).Close(clienturlctx.WithClientURL(closeCtx, di.clientURL), request.GetConnection(), opts...)
```

Comment:
Do we need this `Close`? We didn't make a Request yet.

Reply:
I believe you are half right.  We *may* not have made a Request yet (or we may have).  I believe https://github.com/networkservicemesh/sdk/pull/1083/commits/cc5e48fbeabec05f3e3615c730969b4d7732f348 resolves those two and handles them correctly :)

Reply:
Not actually, I mean that we didn't make a Request to the next at this attempt. I suppose that we should close next in such case:
1. Request next.
2. Fail internal.
3. Close next.
4. Return error.

Because here we will have some update done for the right part of the chain and not recorded in the left part of the chain.

Here we just have the following:
1. Fail internal.
2. Close next.
3. Return error.

Do we actually need this Close? We will either have a new Request or timeout Close to perform a needed cleanup in such case.

Reply:
You are correct, we actually don't need this Close.  Remove it in https://github.com/networkservicemesh/sdk/pull/1083/commits/749c709251a67e2cfd6681f99fecfee5ab2566d4